tpm2/src/util/nsl_util.c

```diff
@@ -0,0 +1,983 @@
+  index += nonce_a_len;
+  memcpy(index, nonce_b, nonce_b_len);
+
+  // Setup the message digest context.
+  //const EVP_MD *type = EVP_sha3_512();
+  const EVP_MD *type = EVP_shake256();
+
+  if (NULL == type)
+  {
+    //kmyth_log(LOG_ERR, "Failed to obtain the SHA512 MD.");
+    kmyth_log(LOG_ERR, "Failed to obtain the SHAKE-256 MD.");
+    free(nonces);
+    return 1;
+  }
+  EVP_MD_CTX *ctx = EVP_MD_CTX_new();
+
+  if (NULL == ctx)
+  {
+    kmyth_log(LOG_ERR, "Failed to create the MD context.");
+    free(nonces);
+    return 1;
+  }
+
+  // Initialize the context and load in the nonces to generate the session
+  // key.
+  int result = EVP_DigestInit_ex(ctx, type, NULL);
+
+  if (0 == result)
+  {
+    kmyth_log(LOG_ERR, "Failed to initialize the MD context.");
+    EVP_MD_CTX_free(ctx);
+    free(nonces);
+    return 1;
+  }
+
+  result = EVP_DigestUpdate(ctx, nonces, len);
+  if (0 == result)
+  {
+    kmyth_log(LOG_ERR, "Failed to update the MD context.");
+    EVP_MD_CTX_free(ctx);
+    free(nonces);
+    return 1;
+  }
+
+  free(nonces);
+
+  *key = calloc(EVP_MAX_MD_SIZE, sizeof(unsigned char));
+  *key_len = EVP_MAX_MD_SIZE;
+
+  result = EVP_DigestFinal_ex(ctx, *key, (unsigned int *) key_len);
+  if (0 == result)
+  {
+    kmyth_log(LOG_ERR, "Failed to finalize the MD context.");
+    EVP_MD_CTX_free(ctx);
+    free(key);
+    *key = NULL;
+    *key_len = 0;
+    return 1;
+  }
+
+  EVP_MD_CTX_free(ctx);
+
+  //if (*key_len != 64)
+  if (*key_len != 32)
+  {
+    kmyth_log(LOG_ERR, "The generated key length must be 32, not %zd.",
+              *key_len);
+    free(key);
+    *key = NULL;
+    *key_len = 0;
+    return 1;
+  }
+
+  // Only use the top half of the generated key.
+  //*key_len = *key_len / 2;
+
+  return 0;
+}
+
+//
+// generate_nonce()
+//
+int generate_nonce(size_t desired_min_nonce_len, unsigned char **nonce,
+                   size_t *nonce_len)
+{
+  size_t size = 1;
+
+  while ((size * sizeof(int)) < desired_min_nonce_len)
+  {
+    size += 1;
+  }
+
+  *nonce_len = size * sizeof(int);
+  unsigned char *buffer = calloc(size, sizeof(int));
+  unsigned char *index = buffer;
+
+  for (size_t i = 0; i < size; i++)
+  {
+    *index = (unsigned char) rand();
+    index += sizeof(int);
+  }
+
+  *nonce = buffer;
+  return 0;
+}
+
+//
+// negotiate_client_session_key()
+//
+int negotiate_client_session_key(int socket_fd,
+                                 EVP_PKEY_CTX * public_key_ctx,
+                                 EVP_PKEY_CTX * private_key_ctx,
+                                 unsigned char *id, size_t id_len,
+                                 unsigned char *expected_id,
+                                 size_t expected_id_len,
+                                 unsigned char **session_key,
+                                 size_t *session_key_len)
+{
+  // Generate nonce A
+  unsigned char *nonce_a = NULL;
+  size_t nonce_a_len = 0;
+
+  int result = generate_nonce(32, &nonce_a, &nonce_a_len);
```

Comment:
Rather than hard-coding the 32-byte nonce length, it may be more readable and configurable if we defined a macro for this value and used that instead.

Reply:
Done